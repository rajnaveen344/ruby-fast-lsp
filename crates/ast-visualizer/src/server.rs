use actix_web::http::header::ContentType;
use actix_web::{web, HttpResponse, Responder, Result};
use ruby_prism::parse;
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::Path;
use std::sync::Mutex;

use crate::ast::convert_node_to_ast;
use crate::utils::offset_to_position;

// Request structure for parsing Ruby code
#[derive(Deserialize, Serialize)]
pub struct ParseRequest {
    pub code: String,
}

// State to track request count
pub struct AppState {
    pub request_count: Mutex<usize>,
}

// Parse Ruby code and return AST as JSON
pub async fn parse_ruby(
    data: web::Json<ParseRequest>,
    state: web::Data<AppState>,
) -> impl Responder {
    // Increment request count - handle poisoned mutex gracefully
    let count_result = state.request_count.lock();
    let request_num = match count_result {
        Ok(mut count) => {
            *count += 1;
            *count
        }
        Err(poisoned) => {
            // If the mutex is poisoned, recover the lock and continue
            let mut count = poisoned.into_inner();
            *count += 1;
            *count
        }
    };
    println!("Request #{}: Parsing Ruby code", request_num);

    // Parse Ruby code
    let code = data.code.clone();

    // Check input size - limit to 500KB to prevent memory issues
    const MAX_INPUT_SIZE: usize = 500 * 1024; // 500KB
    if code.len() > MAX_INPUT_SIZE {
        return HttpResponse::BadRequest().json(serde_json::json!({
            "error": "Input too large",
            "message": format!("Input size ({} bytes) exceeds maximum allowed size ({} bytes)", code.len(), MAX_INPUT_SIZE)
        }));
    }

    // Validate input content
    // Check for non-UTF8 characters or other problematic content
    if code.contains("\0") {
        return HttpResponse::BadRequest().json(serde_json::json!({
            "error": "Invalid input",
            "message": "Input contains null bytes which are not allowed"
        }));
    }

    // Check for extremely long lines that might cause issues
    const MAX_LINE_LENGTH: usize = 10000; // 10K characters per line max
    if code.lines().any(|line| line.len() > MAX_LINE_LENGTH) {
        return HttpResponse::BadRequest().json(serde_json::json!({
            "error": "Invalid input",
            "message": format!("Input contains lines exceeding the maximum allowed length of {} characters", MAX_LINE_LENGTH)
        }));
    }

    // Safely parse the code with error handling
    let parse_result = match std::panic::catch_unwind(|| {
        // Use a more defensive approach to parsing
        parse(code.as_bytes())
    }) {
        Ok(result) => result,
        Err(_) => {
            return HttpResponse::InternalServerError().json(serde_json::json!({
                "error": "Parsing error",
                "message": "Failed to parse the input. The code may be too complex or contain unsupported constructs."
            }));
        }
    };

    // Check for syntax errors
    let errors_count = parse_result.errors().count();

    // If there are too many errors, reject the input
    const MAX_ERRORS: usize = 50;
    if errors_count > MAX_ERRORS {
        return HttpResponse::BadRequest().json(serde_json::json!({
            "error": "Too many syntax errors",
            "message": format!("The input contains {} syntax errors, which exceeds the maximum allowed ({})", errors_count, MAX_ERRORS)
        }));
    }

    // If there are some errors, return them in a structured format
    if errors_count > 0 {
        // Create a structured error response that can be visualized
        let error_nodes: Vec<serde_json::Value> = parse_result
            .errors()
            .map(|e| {
                let location = e.location();
                let message = format!("{:?}", e);

                // Get location information
                let start_offset = location.start_offset();
                let _end_offset = location.end_offset(); // We could use this for error highlighting

                // Calculate line and column from the code string
                let (start_line, start_column) = offset_to_position(&code, start_offset);

                serde_json::json!({
                    "type": "SyntaxErrorNode",
                    "message": message,
                    "line": start_line,
                    "column": start_column
                })
            })
            .collect();

        // Return a special AST structure for errors
        return HttpResponse::Ok().json(serde_json::json!({
            "type": "ErrorTreeNode",
            "name": "Syntax Errors",
            "children": error_nodes
        }));
    }

    // Convert AST to JSON-serializable structure with error handling
    let ast_result = std::panic::catch_unwind(|| convert_node_to_ast(&parse_result.node()));

    match ast_result {
        Ok(ast) => HttpResponse::Ok().json(ast),
        Err(_) => HttpResponse::InternalServerError().json(serde_json::json!({
            "error": "AST processing error",
            "message": "Failed to process the AST. The input may be too complex or contain unsupported constructs."
        }))
    }
}

// Health check endpoint
pub async fn health_check(state: web::Data<AppState>) -> impl Responder {
    // Handle poisoned mutex gracefully
    let count_result = state.request_count.lock();
    let count = match count_result {
        Ok(count) => *count,
        Err(poisoned) => {
            // If the mutex is poisoned, recover the lock and continue
            *poisoned.into_inner()
        }
    };

    HttpResponse::Ok().json(serde_json::json!({
        "status": "ok",
        "requests_processed": count
    }))
}

// Serve the HTML page
pub async fn index() -> Result<HttpResponse> {
    // Path to the HTML file
    let html_path = Path::new("static/index.html");

    // Check if the file exists in the current directory
    if html_path.exists() {
        let html_content = fs::read_to_string(html_path).map_err(|e| {
            eprintln!("Error reading HTML file: {}", e);
            actix_web::error::ErrorInternalServerError("Error reading HTML file")
        })?;

        Ok(HttpResponse::Ok()
            .content_type(ContentType::html())
            .body(html_content))
    } else {
        // Try to find the file relative to the crate directory
        let crate_html_path = Path::new("crates/ast-visualizer/static/index.html");

        if crate_html_path.exists() {
            let html_content = fs::read_to_string(crate_html_path).map_err(|e| {
                eprintln!("Error reading HTML file: {}", e);
                actix_web::error::ErrorInternalServerError("Error reading HTML file")
            })?;

            Ok(HttpResponse::Ok()
                .content_type(ContentType::html())
                .body(html_content))
        } else {
            eprintln!(
                "HTML file not found at {:?} or {:?}",
                html_path, crate_html_path
            );
            Ok(HttpResponse::NotFound().body("HTML file not found"))
        }
    }
}
