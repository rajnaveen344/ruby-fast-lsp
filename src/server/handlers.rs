use anyhow::Result;
use log::{info, warn};
use lsp_types::{
    CompletionResponse, Hover, HoverContents,
    MarkupContent, MarkupKind, Position, Range,
};
use tower_lsp::jsonrpc::Result as LspResult;

use crate::analysis::RubyAnalyzer;
use crate::parser::{document::RubyDocument, RubyParser};

#[derive(Clone)]
pub struct RubyLspHandlers {
    parser: Option<RubyParser>,
    analyzer: RubyAnalyzer,
}

impl RubyLspHandlers {
    pub fn new() -> Result<Self> {
        let parser = match RubyParser::new() {
            Ok(parser) => Some(parser),
            Err(e) => {
                warn!("Failed to initialize Ruby parser: {}", e);
                None
            }
        };
        
        Ok(Self {
            parser,
            analyzer: RubyAnalyzer::new(),
        })
    }

    pub fn handle_hover(&self, document: &RubyDocument, position: Position) -> Option<Hover> {
        info!("Handling hover request at position {:?}", position);
        
        let tree = self.parse_document(document);
        
        // Get hover information
        let hover_info = self
            .analyzer
            .get_hover_info(tree.as_ref(), document.get_content(), position)
            .unwrap_or_else(|| "No information available".to_string());
        
        Some(Hover {
            contents: HoverContents::Markup(MarkupContent {
                kind: MarkupKind::Markdown,
                value: hover_info,
            }),
            range: None,
        })
    }

    pub fn handle_completion(
        &self,
        document: &RubyDocument,
        position: Position,
    ) -> CompletionResponse {
        info!("Handling completion request at position {:?}", position);
        
        let tree = self.parse_document(document);
        
        // Get completions from the analyzer
        let items = self.analyzer.get_completions(tree.as_ref(), document.get_content(), position);
        
        CompletionResponse::Array(items)
    }

    pub fn handle_definition(
        &self,
        document: &RubyDocument,
        position: Position,
    ) -> Option<Range> {
        info!("Handling definition request at position {:?}", position);
        
        let tree = self.parse_document(document);
        
        // Find definition
        self.analyzer.find_definition(tree.as_ref(), document.get_content(), position)
    }
    
    // Helper method to parse a document and return the tree
    fn parse_document(&self, document: &RubyDocument) -> Option<tree_sitter::Tree> {
        // If parser is not available, return None
        let parser = match &self.parser {
            Some(parser) => parser,
            None => {
                warn!("Parser not available for document parsing");
                return None;
            }
        };
        
        // Try to parse the document
        match parser.parse(document.get_content()) {
            Some(tree) => {
                info!("Document parsed successfully");
                Some(tree)
            },
            None => {
                warn!("Failed to parse document");
                None
            }
        }
    }
}
