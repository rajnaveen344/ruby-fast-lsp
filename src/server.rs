use crate::analyzer_prism::visitors::index_visitor::IndexVisitor;
use crate::handlers::{notification, request};
use crate::indexer::index::RubyIndex;
use crate::types::ruby_document::RubyDocument;
use anyhow::Result;
use log::{debug, info};
use lsp_types::{
    CompletionParams, CompletionResponse, DidChangeTextDocumentParams, DidCloseTextDocumentParams,
    DidOpenTextDocumentParams, GotoDefinitionParams, GotoDefinitionResponse, InitializeParams,
    InitializeResult, InitializedParams, InlayHintParams, Location, ReferenceParams,
    SemanticTokensParams, SemanticTokensResult, Url,
};
use ruby_prism::Visit;
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::Instant;
use tower_lsp::jsonrpc::Result as LspResult;
use tower_lsp::{Client, LanguageServer};

#[derive(Clone)]
pub struct RubyLanguageServer {
    pub client: Option<Client>,
    pub index: Arc<Mutex<RubyIndex>>,
    pub docs: Arc<Mutex<HashMap<Url, RubyDocument>>>,
}

impl RubyLanguageServer {
    pub fn new(client: Client) -> Result<Self> {
        let index = RubyIndex::new();
        Ok(Self {
            client: Some(client),
            index: Arc::new(Mutex::new(index)),
            docs: Arc::new(Mutex::new(HashMap::new())),
        })
    }

    pub fn index(&self) -> Arc<Mutex<RubyIndex>> {
        self.index.clone()
    }

    pub fn get_doc(&self, uri: &Url) -> Option<RubyDocument> {
        self.docs.lock().unwrap().get(uri).cloned()
    }

    pub fn process_file(&mut self, uri: Url, content: &str) -> Result<(), String> {
        self.index.lock().unwrap().remove_entries_for_uri(&uri);

        // Create or update document in the docs HashMap
        let document = RubyDocument::new(uri.clone(), content.to_string(), 0);
        self.docs.lock().unwrap().insert(uri.clone(), document);

        let parse_result = ruby_prism::parse(content.as_bytes());
        let node = parse_result.node();
        let mut visitor = IndexVisitor::new(self, uri.clone());

        visitor.visit(&node);

        // Persist mutations made by the visitor back to the server's document store
        // TODO: This is a temporary fix. We should be able to mutate the document in place
        //       using docs: Arc<Mutex<HashMap<Url, Arc<Mutex<RubyDocument>>>>>
        self.docs
            .lock()
            .unwrap()
            .insert(uri.clone(), visitor.document.clone());

        debug!("Processed file: {}", uri);
        Ok(())
    }
}

impl Default for RubyLanguageServer {
    fn default() -> Self {
        RubyLanguageServer {
            client: None,
            index: Arc::new(Mutex::new(RubyIndex::new())),
            docs: Arc::new(Mutex::new(HashMap::new())),
        }
    }
}

#[tower_lsp::async_trait]
impl LanguageServer for RubyLanguageServer {
    async fn initialize(&self, params: InitializeParams) -> LspResult<InitializeResult> {
        notification::handle_initialize(self, params).await
    }

    async fn initialized(&self, params: InitializedParams) {
        notification::handle_initialized(self, params).await
    }

    async fn shutdown(&self) -> LspResult<()> {
        notification::handle_shutdown(self).await
    }

    async fn did_open(&self, params: DidOpenTextDocumentParams) {
        info!("Document opened: {}", params.text_document.uri.path());
        let start_time = Instant::now();
        notification::handle_did_open(self, params).await;
        info!(
            "[PERF] Document open handler completed in {:?}",
            start_time.elapsed()
        );
    }

    async fn did_change(&self, params: DidChangeTextDocumentParams) {
        info!("Document changed: {}", params.text_document.uri.path());
        let start_time = Instant::now();
        notification::handle_did_change(self, params).await;
        info!(
            "[PERF] Document change handler completed in {:?}",
            start_time.elapsed()
        );
    }

    async fn did_close(&self, params: DidCloseTextDocumentParams) {
        info!("Document closed: {}", params.text_document.uri.path());
        let start_time = Instant::now();
        notification::handle_did_close(self, params).await;
        info!(
            "[PERF] Document close handler completed in {:?}",
            start_time.elapsed()
        );
    }

    async fn goto_definition(
        &self,
        params: GotoDefinitionParams,
    ) -> LspResult<Option<GotoDefinitionResponse>> {
        info!(
            "Goto definition request received for {:?}",
            params
                .text_document_position_params
                .text_document
                .uri
                .path()
        );
        let start_time = Instant::now();
        let result = request::handle_goto_definition(self, params).await;

        info!(
            "[PERF] Goto definition completed in {:?}",
            start_time.elapsed()
        );

        result
    }

    async fn references(&self, params: ReferenceParams) -> LspResult<Option<Vec<Location>>> {
        info!(
            "References request received for {:?}",
            params.text_document_position.text_document.uri.path()
        );
        let start_time = Instant::now();
        let result = request::handle_references(self, params).await;

        info!("[PERF] References completed in {:?}", start_time.elapsed());

        result
    }

    async fn semantic_tokens_full(
        &self,
        params: SemanticTokensParams,
    ) -> LspResult<Option<SemanticTokensResult>> {
        info!(
            "Semantic tokens request received for {:?}",
            params.text_document.uri.path()
        );
        let start_time = Instant::now();
        let result = request::handle_semantic_tokens_full(self, params).await;

        info!(
            "[PERF] Semantic tokens completed in {:?}",
            start_time.elapsed()
        );

        result
    }

    async fn inlay_hint(
        &self,
        params: InlayHintParams,
    ) -> LspResult<Option<Vec<lsp_types::InlayHint>>> {
        info!(
            "Inlay hint request received for {:?}",
            params.text_document.uri.path()
        );

        let start_time = Instant::now();
        let result = request::handle_inlay_hints(self, params).await;

        info!("[PERF] Inlay hint completed in {:?}", start_time.elapsed());

        result
    }

    async fn completion(&self, params: CompletionParams) -> LspResult<Option<CompletionResponse>> {
        info!(
            "Completion request received for {:?}",
            params.text_document_position.text_document.uri.path()
        );
        let start_time = Instant::now();
        let result = request::handle_completion(self, params).await;

        info!("[PERF] Completion completed in {:?}", start_time.elapsed());

        result
    }
}
