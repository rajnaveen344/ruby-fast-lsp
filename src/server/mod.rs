pub mod handlers;

use anyhow::Result;
use dashmap::DashMap;
use log::{info, warn};
use lsp_types::*;
use tower_lsp::jsonrpc::Result as LspResult;
use tower_lsp::{Client, LanguageServer};

use crate::parser::document::RubyDocument;
use self::handlers::RubyLspHandlers;
use std::sync::{Arc, Mutex};

pub struct RubyLanguageServer {
    pub client: Client,
    pub document_map: DashMap<Url, RubyDocument>,
    handlers: Option<Arc<Mutex<RubyLspHandlers>>>,
}

impl RubyLanguageServer {
    pub fn new(client: Client) -> Result<Self> {
        let handlers = Some(Arc::new(Mutex::new(RubyLspHandlers::new()?)));
        
        Ok(Self {
            client,
            document_map: DashMap::new(),
            handlers,
        })
    }
    
    pub fn new_fallback(client: Client) -> Self {
        warn!("Creating fallback Ruby LSP server with limited functionality");
        
        Self {
            client,
            document_map: DashMap::new(),
            handlers: None,
        }
    }
}

#[tower_lsp::async_trait]
impl LanguageServer for RubyLanguageServer {
    async fn initialize(&self, _: InitializeParams) -> LspResult<InitializeResult> {
        info!("Initializing Ruby LSP server");
        
        let capabilities = ServerCapabilities {
            text_document_sync: Some(TextDocumentSyncCapability::Kind(
                TextDocumentSyncKind::INCREMENTAL,
            )),
            hover_provider: Some(HoverProviderCapability::Simple(true)),
            completion_provider: Some(CompletionOptions {
                resolve_provider: Some(false),
                trigger_characters: Some(vec![".".to_string(), "::".to_string()]),
                work_done_progress_options: Default::default(),
                all_commit_characters: None,
                completion_item: None,
            }),
            definition_provider: Some(OneOf::Left(true)),
            // Add more capabilities as you implement them
            ..ServerCapabilities::default()
        };
        
        Ok(InitializeResult {
            capabilities,
            server_info: Some(ServerInfo {
                name: "ruby-fast-lsp".to_string(),
                version: Some(env!("CARGO_PKG_VERSION").to_string()),
            }),
        })
    }

    async fn initialized(&self, _: InitializedParams) {
        info!("Ruby LSP server initialized!");
        
        let message = if self.handlers.is_some() {
            "Ruby LSP server initialized with full functionality!"
        } else {
            "Ruby LSP server initialized with limited functionality. Some features may not work correctly."
        };
        
        self.client
            .log_message(MessageType::INFO, message)
            .await;
    }

    async fn shutdown(&self) -> LspResult<()> {
        info!("Shutting down Ruby LSP server");
        Ok(())
    }

    // Document synchronization methods
    async fn did_open(&self, params: DidOpenTextDocumentParams) {
        let uri = params.text_document.uri;
        let text = params.text_document.text;
        let version = params.text_document.version;
        
        info!("File opened: {}", uri);
        
        self.document_map.insert(
            uri.clone(),
            RubyDocument {
                content: text.clone(),
                version,
            },
        );
        
        self.client
            .log_message(MessageType::INFO, &format!("Opened document: {}", uri))
            .await;
    }

    async fn did_change(&self, params: DidChangeTextDocumentParams) {
        let uri = params.text_document.uri;
        let version = params.text_document.version;
        
        info!("File changed: {}", uri);
        
        if let Some(mut doc) = self.document_map.get_mut(&uri) {
            for change in params.content_changes {
                if let Some(_range) = change.range {
                    // Handle incremental updates
                    doc.content = change.text;
                } else {
                    // Full document update
                    doc.content = change.text;
                }
                doc.version = version;
            }
        }
    }

    async fn did_close(&self, params: DidCloseTextDocumentParams) {
        let uri = params.text_document.uri;
        
        info!("File closed: {}", uri);
        
        self.document_map.remove(&uri);
        
        self.client
            .log_message(MessageType::INFO, &format!("Closed document: {}", uri))
            .await;
    }

    // Language features
    async fn hover(&self, params: HoverParams) -> LspResult<Option<Hover>> {
        let uri = params.text_document_position_params.text_document.uri;
        let position = params.text_document_position_params.position;
        
        info!("Hover request at position {:?} in {}", position, uri);
        
        if let Some(handlers) = &self.handlers {
            if let Some(doc) = self.document_map.get(&uri) {
                if let Ok(handlers) = handlers.lock() {
                    return Ok(handlers.handle_hover(&doc, position));
                }
            }
        }
        
        // Fallback hover response
        Ok(Some(Hover {
            contents: HoverContents::Markup(MarkupContent {
                kind: MarkupKind::Markdown,
                value: "Ruby LSP running in fallback mode. Limited functionality available.".to_string(),
            }),
            range: None,
        }))
    }

    async fn completion(&self, params: CompletionParams) -> LspResult<Option<CompletionResponse>> {
        let uri = params.text_document_position.text_document.uri;
        let position = params.text_document_position.position;
        
        info!("Completion request at position {:?} in {}", position, uri);
        
        if let Some(handlers) = &self.handlers {
            if let Some(doc) = self.document_map.get(&uri) {
                if let Ok(handlers) = handlers.lock() {
                    return Ok(Some(handlers.handle_completion(&doc, position)));
                }
            }
        }
        
        // Fallback completion response with basic Ruby keywords
        let items = vec![
            CompletionItem {
                label: "def".to_string(),
                kind: Some(CompletionItemKind::KEYWORD),
                detail: Some("Define a method".to_string()),
                ..CompletionItem::default()
            },
            CompletionItem {
                label: "class".to_string(),
                kind: Some(CompletionItemKind::KEYWORD),
                detail: Some("Define a class".to_string()),
                ..CompletionItem::default()
            },
            CompletionItem {
                label: "module".to_string(),
                kind: Some(CompletionItemKind::KEYWORD),
                detail: Some("Define a module".to_string()),
                ..CompletionItem::default()
            },
        ];
        
        Ok(Some(CompletionResponse::Array(items)))
    }
    
    async fn goto_definition(
        &self,
        params: GotoDefinitionParams,
    ) -> LspResult<Option<GotoDefinitionResponse>> {
        let uri = params.text_document_position_params.text_document.uri;
        let position = params.text_document_position_params.position;
        
        info!("Definition request at position {:?} in {}", position, uri);
        
        if let Some(handlers) = &self.handlers {
            if let Some(doc) = self.document_map.get(&uri) {
                if let Ok(handlers) = handlers.lock() {
                    if let Some(range) = handlers.handle_definition(&doc, position) {
                        return Ok(Some(GotoDefinitionResponse::Scalar(Location {
                            uri: uri.clone(),
                            range,
                        })));
                    }
                }
            }
        }
        
        Ok(None)
    }
}
