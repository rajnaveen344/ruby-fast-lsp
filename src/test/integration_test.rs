//! Integration test harness – see `docs/integration_test_plan.md`
//!
//! This file purposely **replaces** the previous ad-hoc integration tests with a reusable
//! and extensible test harness.  The harness makes it straightforward to add new fixtures
//! and coverage as described in the integration test plan.

use std::path::{Path, PathBuf};
use std::sync::Once;

use serde_json::Value;
use tower_lsp::lsp_types::{DidOpenTextDocumentParams, InitializeParams, TextDocumentItem, Url};
use tower_lsp::LanguageServer;

use crate::server::RubyLanguageServer;

/*----------------------------------------------------------------------
 Logger
----------------------------------------------------------------------*/

fn init_logger() {
    static INIT: Once = Once::new();
    INIT.call_once(|| {
        // Respect RUST_LOG env var but default to info for the test binary.
        env_logger::Builder::from_env(env_logger::Env::default().default_filter_or("debug"))
            .is_test(true)
            .init();
    });
}

/*----------------------------------------------------------------------
 Fixture helpers
----------------------------------------------------------------------*/

/// Absolute path to the root `fixtures/` directory.
fn fixture_root() -> PathBuf {
    // `CARGO_MANIFEST_DIR` resolves to the crate root at compile time.
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("src")
        .join("test")
        .join("fixtures")
}

/// Convert a [`Path`] to an [`Url`].  Panics if the conversion fails.
fn path_to_uri(path: &Path) -> Url {
    Url::from_file_path(path).expect("Failed to convert path to file:// URI")
}

/*----------------------------------------------------------------------
 TestHarness
----------------------------------------------------------------------*/

/// Encapsulates a `RubyLanguageServer` instance and helper utilities for
/// loading fixtures and performing common LSP requests.
pub struct TestHarness {
    server: RubyLanguageServer,
}

impl TestHarness {
    /// Create a fresh `RubyLanguageServer` instance and perform the
    /// `initialize` handshake.
    pub async fn new() -> Self {
        init_logger();

        let server = RubyLanguageServer::default();
        let _ = server.initialize(InitializeParams::default()).await;

        Self { server }
    }

    /// Opens `*.rb` files located under `fixtures/<scenario>` so that the
    /// server indexes them as a workspace.
    ///
    /// If `scenario` refers to a single Ruby file (e.g. `some_dir/foo.rb`) that
    /// file alone is opened instead of scanning a directory. This makes it
    /// possible to test language-server behaviour when only one document is
    /// open.
    pub async fn open_fixture_dir(&self, scenario: &str) {
        let root_path = fixture_root().join(scenario);
        assert!(root_path.exists(), "Unknown fixture scenario: {}", scenario);

        // -------------------------------------------------------------
        // Single-file mode – open the requested file and return early
        // -------------------------------------------------------------
        if root_path.is_file() {
            assert!(
                root_path.extension().and_then(|ext| ext.to_str()) == Some("rb"),
                "Expected a .rb file"
            );

            let uri = path_to_uri(&root_path);
            let contents = std::fs::read_to_string(&root_path)
                .unwrap_or_else(|_| panic!("Failed to read {:?}", root_path));

            let params = DidOpenTextDocumentParams {
                text_document: TextDocumentItem {
                    uri,
                    language_id: "ruby".into(),
                    version: 1,
                    text: contents,
                },
            };

            // Fire-and-await so the document is fully indexed before continuing.
            self.server.did_open(params).await;
            return;
        }

        // -------------------------------------------------------------
        // Directory mode – recursively open every Ruby file we find
        // -------------------------------------------------------------
        let mut stack = vec![root_path];
        while let Some(dir) = stack.pop() {
            for entry in std::fs::read_dir(&dir).expect("read_dir failed") {
                let entry = entry.expect("DirEntry failed");
                let path = entry.path();
                if path.is_dir() {
                    stack.push(path);
                } else if path.extension().and_then(|ext| ext.to_str()) == Some("rb") {
                    let uri = path_to_uri(&path);
                    let contents = std::fs::read_to_string(&path)
                        .unwrap_or_else(|_| panic!("Failed to read {:?}", path));

                    let params = DidOpenTextDocumentParams {
                        text_document: TextDocumentItem {
                            uri,
                            language_id: "ruby".into(),
                            version: 1,
                            text: contents,
                        },
                    };

                    // Fire-and-await so the document is fully indexed before continuing.
                    self.server.did_open(params).await;
                }
            }
        }
    }

    /// Borrow the underlying server – useful when calling handlers directly.
    pub fn server(&self) -> &RubyLanguageServer {
        &self.server
    }
}

/*----------------------------------------------------------------------
 Snapshot helpers
----------------------------------------------------------------------*/

/// Capture the reference locations at (`file`, `line`, `char`) and snapshot
/// the JSON array so it is easy to review when behaviour changes.
pub async fn snapshot_references(
    harness: &TestHarness,
    file: &str,
    line: u32,
    character: u32,
    snapshot_name: &str,
) {
    use crate::handlers::request;
    use tower_lsp::lsp_types::{
        PartialResultParams, Position, ReferenceContext, ReferenceParams, TextDocumentIdentifier,
        TextDocumentPositionParams, WorkDoneProgressParams,
    };

    let uri = path_to_uri(&fixture_root().join(file));
    let res_opt = request::handle_references(
        harness.server(),
        ReferenceParams {
            text_document_position: TextDocumentPositionParams {
                text_document: TextDocumentIdentifier { uri: uri.clone() },
                position: Position { line, character },
            },
            work_done_progress_params: WorkDoneProgressParams::default(),
            partial_result_params: PartialResultParams::default(),
            context: ReferenceContext {
                include_declaration: true,
            },
        },
    )
    .await
    .expect("goto references failed");

    let mut value = match res_opt {
        Some(locations) => serde_json::to_value(&locations).unwrap(),
        None => serde_json::json!([]),
    };

    let project_root = std::path::PathBuf::from(std::env::var("CARGO_MANIFEST_DIR").unwrap());
    relativize_uris(&mut value, &project_root);

    insta::assert_json_snapshot!(snapshot_name, value);
}

/// Capture the definition locations at (`file`, `line`, `char`) and snapshot
/// the JSON array so it is easy to review when behaviour changes.
pub async fn snapshot_definitions(
    harness: &TestHarness,
    file: &str,
    line: u32,
    character: u32,
    snapshot_name: &str,
) {
    use crate::handlers::request;
    use tower_lsp::lsp_types::{
        GotoDefinitionParams, PartialResultParams, Position, TextDocumentIdentifier,
        TextDocumentPositionParams, WorkDoneProgressParams,
    };

    let uri = path_to_uri(&fixture_root().join(file));
    let res_opt = request::handle_goto_definition(
        harness.server(),
        GotoDefinitionParams {
            text_document_position_params: TextDocumentPositionParams {
                text_document: TextDocumentIdentifier { uri: uri.clone() },
                position: Position { line, character },
            },
            work_done_progress_params: WorkDoneProgressParams::default(),
            partial_result_params: PartialResultParams::default(),
        },
    )
    .await
    .expect("goto definition failed");

    // Convert the LSP response (if any) into a JSON value so it can be snapshotted.
    // If there is **no** definition then we snapshot an empty JSON array so callers can
    // assert on the absence of definitions without causing a test failure.
    let mut value = match res_opt {
        Some(tower_lsp::lsp_types::GotoDefinitionResponse::Array(loc)) => {
            serde_json::to_value(&loc).unwrap()
        }
        Some(tower_lsp::lsp_types::GotoDefinitionResponse::Scalar(l)) => {
            serde_json::to_value(&vec![l]).unwrap()
        }
        Some(tower_lsp::lsp_types::GotoDefinitionResponse::Link(ls)) => {
            serde_json::to_value(&ls).unwrap()
        }
        None => serde_json::json!([]),
    };

    let project_root = std::path::PathBuf::from(std::env::var("CARGO_MANIFEST_DIR").unwrap());
    relativize_uris(&mut value, &project_root);

    insta::assert_json_snapshot!(snapshot_name, value);
}

/// Replace absolute `file://` URIs inside `value` with a `$PROJECT_ROOT` placeholder so
/// that Insta snapshots are stable across machines.
///
/// * `value` – JSON value that will be mutated in-place.
/// * `project_root` – absolute path to the crate root (usually `env!("CARGO_MANIFEST_DIR")`).
fn relativize_uris(value: &mut Value, project_root: &Path) {
    // Build a canonical `file://` prefix using forward slashes so the check works on all OSes.
    let mut prefix = String::from("file://");
    prefix.push_str(&project_root.display().to_string().replace('\\', "/"));
    if !prefix.ends_with('/') {
        prefix.push('/');
    }

    // Recursive helper – kept private to the function.
    fn walk(v: &mut Value, prefix: &str) {
        match v {
            Value::String(s) if s.starts_with(prefix) => {
                let rel = &s[prefix.len()..];
                *s = format!("file://$PROJECT_ROOT/{}", rel);
            }
            Value::Array(arr) => arr.iter_mut().for_each(|child| walk(child, prefix)),
            Value::Object(map) => map.values_mut().for_each(|child| walk(child, prefix)),
            _ => {}
        }
    }

    walk(value, &prefix);
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::analyzer_prism::visitors::index_visitor::IndexVisitor;
    use crate::indexer::dependency_tracker::DependencyTracker;
    use crate::server::RubyLanguageServer;
    use crate::types::ruby_document::RubyDocument;
    use parking_lot::{Mutex, RwLock};
    use std::path::PathBuf;
    use std::sync::Arc;
    use tower_lsp::lsp_types::Url;

    /// Simply verifies that we can load a fixture directory without panicking.
    #[tokio::test]
    async fn harness_smoke() {
        let harness = TestHarness::new().await;
        harness.open_fixture_dir("goto").await;
        assert!(true);
    }

    #[tokio::test]
    async fn test_dependency_tracking_integration() {
        init_logger();

        // Create a temporary directory for test files
        let temp_dir = tempfile::TempDir::new().unwrap();
        let temp_path = temp_dir.path();
        
        // Create the lib directory and set.rb file
        let lib_dir = temp_path.join("lib");
        std::fs::create_dir_all(&lib_dir).unwrap();
        let set_file = lib_dir.join("set.rb");
        std::fs::write(&set_file, "class Set\n  # Set implementation\nend\n").unwrap();

        // Create a mock server
        let server = RubyLanguageServer::default();

        // Create dependency tracker
        let lib_dirs = vec![lib_dir.clone()];
        let dependency_tracker = Arc::new(Mutex::new(DependencyTracker::new(
            temp_path.to_path_buf(),
            lib_dirs,
        )));

        // Create test file URL
        let test_file = temp_path.join("test_set.rb");
        let uri = Url::from_file_path(&test_file).unwrap();

        // Test content with require_relative
        let content = r#"
require_relative 'lib/set'

class TestClass
  def initialize
    @my_set = Set.new
  end
end
"#;

        // Write the test file to disk
        std::fs::write(&test_file, content).unwrap();

        // Insert a RubyDocument so that IndexVisitor::new can retrieve it
        let doc = RubyDocument::new(uri.clone(), content.to_string(), 0);
        server
            .docs
            .lock()
            .insert(uri.clone(), Arc::new(RwLock::new(doc.clone())));

        // Create IndexVisitor with dependency tracker
        let mut visitor = IndexVisitor::new(&server, uri.clone())
            .with_dependency_tracker(dependency_tracker.clone());

        // Parse and visit the content
        let parse_result = ruby_prism::parse(content.as_bytes());
        let root = parse_result.node();

        use ruby_prism::Visit;
        visitor.visit(&root);

        // Check that the dependency was tracked
        let tracker_guard = dependency_tracker.lock();
        let stats = tracker_guard.get_stats();

        println!("Dependency stats: {:?}", stats);
        assert!(
            stats.total_dependencies > 0,
            "Should have tracked at least one dependency"
        );
    }
}
