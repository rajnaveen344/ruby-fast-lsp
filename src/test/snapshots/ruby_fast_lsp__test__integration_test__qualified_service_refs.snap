---
source: src/test/integration_test.rs
expression: value
---
[
  {
    "range": {
      "end": {
        "character": 42,
        "line": 20
      },
      "start": {
        "character": 8,
        "line": 20
      }
    },
    "uri": "file://$PROJECT_ROOT/src/test/fixtures/qualified_method_call.rb"
  },
  {
    "range": {
      "end": {
        "character": 44,
        "line": 46
      },
      "start": {
        "character": 0,
        "line": 46
      }
    },
    "uri": "file://$PROJECT_ROOT/src/test/fixtures/qualified_method_call.rb"
  },
  {
    "range": {
      "end": {
        "character": 0,
        "line": 50
      },
      "start": {
        "character": 19,
        "line": 49
      }
    },
    "uri": "file://$PROJECT_ROOT/src/test/fixtures/qualified_method_call.rb"
  }
]
